src/components/CardList/CardList.tsx

```diff
@@ -0,0 +1,294 @@
+}
+
+interface CardListProps {
+  searchTerm: string;
+}
+
+function CardList({ searchTerm }: CardListProps): JSX.Element {
+  const [searchParams, setSearchParams] = useSearchParams();
+  const { page } = useParams<{ page: string }>();
+  const currentPage = parseInt(page || searchParams.get('page') || '1', 10);
+  const pokemonId = searchParams.get('pokemonId');
+  const itemsPerPage = 24;
+  const [pokemonList, setPokemonList] = useState<Pokemon[]>([]);
+  const [selectedPokemon, setSelectedPokemon] = useState<PokemonDetails | null>(
+    null
+  );
+  const [isLoading, setIsLoading] = useState<boolean>(false);
+  const [isDetailsLoading, setIsDetailsLoading] = useState<boolean>(false);
+  const [error, setError] = useState<string | null>(null);
+  const [detailsError, setDetailsError] = useState<string | null>(null);
+  const [totalPages, setTotalPages] = useState<number>(1);
+
+  useEffect(() => {
+    async function fetchPokemon(term: string) {
+      setIsLoading(true);
+      setError(null);
+      try {
+        let pokemonDetails: Pokemon[] = [];
+        if (term) {
+          const response = await fetch(
+            `https://pokeapi.co/api/v2/pokemon/${term.toLowerCase()}`
+          );
+          if (!response.ok) {
+            throw new Error(
+              response.status === 404
+                ? 'No Pokémon found'
+                : `HTTP error! Status: ${response.status}`
+            );
+          }
+          const data = await response.json();
+          pokemonDetails = [
+            { ...data, url: `https://pokeapi.co/api/v2/pokemon/${data.id}/` },
+          ];
+          setTotalPages(1);
+          if (pokemonId) {
+            setSearchParams({ page: currentPage.toString() });
+          }
+        } else {
+          const offset = (currentPage - 1) * itemsPerPage;
+          const response = await fetch(
+            `https://pokeapi.co/api/v2/pokemon?limit=${itemsPerPage}&offset=${offset}`
+          );
+          if (!response.ok) {
+            throw new Error(`HTTP error! Status: ${response.status}`);
+          }
+          const data = await response.json();
+          pokemonDetails = await Promise.all(
+            data.results.map(async (summary: PokemonSummary) => {
+              const detailResponse = await fetch(summary.url);
+              if (!detailResponse.ok) {
+                throw new Error(`HTTP error! Status: ${detailResponse.status}`);
+              }
+              return await detailResponse.json();
+            })
+          );
+          setTotalPages(Math.ceil(data.count / itemsPerPage));
+        }
+        setPokemonList(pokemonDetails);
+        setIsLoading(false);
+      } catch (error) {
+        setError(error instanceof Error ? error.message : 'An error occurred');
+        setIsLoading(false);
+      }
+    }
+
+    fetchPokemon(searchTerm);
+  }, [searchTerm, currentPage, pokemonId, setSearchParams]);
+
+  useEffect(() => {
+    async function fetchPokemonDetails() {
```

Comment:
Consider moving data fetching logic into a separate function or custom hook to keep the component cleaner and make the code easier to test and reuse.